README.md
# Qpid PMDA
[![Build Status](http://img.shields.io/travis/pcolby/pcp-pmda-qpid.svg)](https://travis-ci.org/pcolby/pcp-pmda-qpid)
[![Github Release](http://img.shields.io/github/release/pcolby/pcp-pmda-qpid.svg)](https://github.com/pcolby/pcp-pmda-qpid/releases/latest)
[![Apache License](http://img.shields.io/badge/license-APACHE2-blue.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)


Qpid PMDA is a free, open source, [PCP](http://oss.sgi.com/projects/pcp/
"Performance Co-Pilot") Performance Metrics Domain Agent for [Apache Qpid](
http://qpid.apache.org/). It exposes performance metrics from Qpid message
brokers via [QMF](http://qpid.apache.org/components/qmf/index.html "Qpid
Management Framework"), to Performance Co-Pilot (PCP) using the [PMDA++](
https://github.com/pcolby/pcp-pmda-cpp) library.

```
      PCP      |   libpcp    |        | pmdaqpid-qmf1 | QMF1 |     Qpid
  Monitoring   |-------------| PMDA++ |---------------+------|   Messaging
Infrastructure | libpcp_pmda |        | pmdaqpid-qmf2 | QMF2 | Infrastructure
```

## Requirements

* [Boost C++ Libraries](http://www.boost.org/) 1.32+
* [PMDA++](https://github.com/pcolby/pcp-pmda-cpp) 0.4.0+
* [Apache Qpid](http://qpid.apache.org/)

## Building & Installing
1. Install the pre-requisites - Boost, Apache Qpid, [PMDA++](https://github.com/pcolby/pcp-pmda-cpp).
2. `mkdir build && cd build`
3. `cmake <path-to-code> && make && sudo make install`
4. ``cd `pmconfig PCP_PMDAS_DIR | cut -b15-`/qpid && ./Install``

Alternatively, use [rpmbuild](package/rpm).

## Contact

Issues / suggestions can be reported via GitHub's [issue tracker](
https://github.com/pcolby/pcp-pmda-qpid/issues) (pull requests are very
welcome).

The [PMDA++ Google Group](http://groups.google.com/group/pcp-pmda-cpp/) is also
used for general discussion, questions, comments, suggestions, etc for this
project.  Email pcp-pmda-cpp+subscribe@googlegroups.com to subscribe.

## License

Available under the [Apache License, Version 2.0](
http://apache.org/licenses/LICENSE-2.0.html).


CHANGELOG.md
## Qpid PMDA Changelog

### 0.2.5 (_unreleased_)
Bug fixes:
- `QpidPmdaQmf1::nonPmdaMode` not initialised in constructor
  ([e8d6093](../../commit/e8d6093a0d662f89585adca4217f89ee3cf5eb41))

### 0.2.4 (2014-05-15)
Features:
- use `pmda::cache::*` methods ([88da366](
  ../../commit/88da366a5b82415a25bc7f639778f0b84a737614)) -
  requires PMDA++ 0.4.0 or later.

### 0.2.3 (2014-05-10)
Bug fixes:
- verbose metrics descriptions fallback ([#19](../../issues/19))

### 0.2.2 (2014-03-21)
Bug fixes:
- `DESTDIR` support for exported files ([#8](../../issues/8))
- added missing metrics descriptions ([#11](../../issues/11))

### 0.2.1 (2014-03-05)
Features:
- extensive code documentation.

Bug fixes:
- memory leak fix ([#9](../../issues/9))

### 0.2.0 (2014-02-18)
Features:
- code refactor for combined QMFv1 and QMFv2 support in one binary.

### 0.1.1 (2014-02-16)
Features:
- NSS CLI support ([#5](../../issues/5))
- Qpid log output ([#7](../../issues/7))

Bug fixes:
- `--transport` command line option implemented ([39bb3d7](
  ../../commit/39bb3d7f72a30d49b8a0f1efa7e970169e8373b8))

### 0.1.0 (2014-02-15)
- initial release.
- QMFv1 support only.


src/qmf1/ConsoleUtils.cpp
/*
 * Copyright 2013-2014 Paul Colby
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file
 * @brief Defines the ConsoleUtils class.
 */

#include "ConsoleUtils.h"

#include <qpid/console/Value.h>

#include <boost/lexical_cast.hpp>

/**
 * @brief Get a standardised name for a QMF object.
 *
 * This function checks a QMF object for an appropriate "name" attribute, and if
 * found, returns it verbatim.  If the object has no "name" attribute, and
 * allowNodeName is \c true, and the object has a "nodeName" attribute, then
 * that attribute's value will be returned verbatim.  Otherwise, and empty
 * string is returned.
 *
 * @param object        Object to fetch the name of.
 * @param allowNodeName If \c true, and \a object has no name attribute, the
 *                      nodeName attribute will be considered also.
 *
 * @return A name for \a object, or an empty string if no name could be found.
 */
std::string ConsoleUtils::getName(const qpid::console::Object &object,
                                  const bool allowNodeName)
{
    const qpid::console::Object::AttributeMap &attributes = object.getAttributes();
    qpid::console::Object::AttributeMap::const_iterator name = attributes.find("name");
    if ((name == attributes.end()) && (allowNodeName)) {
        name = attributes.find("nodeName");
    }
    return (name == attributes.end()) ? std::string() : name->second->asString();
}

/**
 * @brief Get the schema type of a QMF object.
 *
 * @param object QMF object to fetch the type of.
 *
 * @return One of the ConsoleUtils::ObjectSchemaType enumeration values.
 */
ConsoleUtils::ObjectSchemaType ConsoleUtils::getType(const qpid::console::Object &object)
{
    return getType(object.getClassKey());
}

/**
 * @brief Get the schema type of a QMF class key.
 *
 * @param classKey QMF class key to fetch the type of.
 *
 * @return One of the ConsoleUtils::ObjectSchemaType enumeration values.
 */
ConsoleUtils::ObjectSchemaType ConsoleUtils::getType(const qpid::console::ClassKey &classKey)
{
    if (classKey.getPackageName() == "org.apache.qpid.broker") {
        const std::string &className = classKey.getClassName();
        if (className == "broker") return Broker;
        if (className == "queue")  return Queue;
        if (className == "system") return System;
    }
    return Other;
}

/**
 * @brief Convert a QMF type code to a human-readable string.
 *
 * This function is only used to make log messages more friendly.
 *
 * @param typeCode QMF type code to convert to string.
 *
 * @return A human-readable version of \a typeCode.
 */
std::string ConsoleUtils::qmfTypeCodeToString(const uint8_t typeCode)
{
    // See Qpid's cpp/include/qmf/engine/Typecode.h
    switch (typeCode) {
        case 1: return "UINT8";
        case 2: return "UINT16";
        case 3: return "UINT32";
        case 4: return "UINT64";
      //case 5: // There is no type 5.
        case 6: return "SSTR";
        case 7: return "LSTR";
        case 8: return "ABSTIME";
        case 9: return "DELTATIME";
        case 10: return "REF";
        case 11: return "BOOL";
        case 12: return "FLOAT";
        case 13: return "DOUBLE";
        case 14: return "UUID";
        case 15: return "MAP";
        case 16: return "INT8";
        case 17: return "INT16";
        case 18: return "INT32";
        case 19: return "INT64";
        case 20: return "OBJECT";
        case 21: return "LIST";
        case 22: return "ARRAY";
        default:
            return "unknown (" + boost::lexical_cast<std::string>(typeCode) + ')';
    }
}

/**
 * @brief Convert a QMF class to a canonical string for logging.
 *
 * @param classKey QMF class key to convert to string.
 *
 * @return String representation of \a classKey, suitable for logging.
 */
std::string ConsoleUtils::toString(const qpid::console::ClassKey &classKey)
{
    return classKey.getPackageName() + ':' + classKey.getClassName();
}

/**
 * @brief Convert a QMF object to a canonical string for logging.
 *
 * @param object             QMF object to conver to string.
 * @param includePackageName If \c true, the object's QMF package name will be
 *                           included in the returned string.
 *
 * @return String representation of \a object, suitable for logging.
 */
std::string ConsoleUtils::toString(const qpid::console::Object &object,
                                   bool includePackageName)
{
    const qpid::console::ClassKey &classKey = object.getClassKey();
    return (includePackageName ? toString(classKey) : classKey.getClassName()) +
           " '" + getName(object) + "' (" + toString(object.getObjectId()) + ')';
}

/**
 * @brief Convert a QMF object ID to a canonical string for logging.
 *
 * @param id QMF object ID to convert to string.
 *
 * @return String representation of \a id, suitable for logging.
 */
std::string ConsoleUtils::toString(const qpid::console::ObjectId &id)
{
    std::ostringstream stream; // Take advantage of the qpid::console::ObjectId
    stream << id;              // class' built-in stream operator.
    return stream.str();
}

/**
 * @brief Convert a QMF property schema to a canonical string for logging.
 *
 * @param property QMF property schema to convert to string.
 *
 * @return String representation of \a property, suitable for logging.
 */
std::string ConsoleUtils::toString(const qpid::console::SchemaProperty &property)
{
    return property.name + ':' + qmfTypeCodeToString(property.typeCode) + ':' +
           property.unit + ':' + property.desc;
}

/**
 * @brief Convert a QMF statistic schema to a canonical string for logging.
 *
 * @param statistic QMF statistic schema to convert to string.
 *
 * @return String representation of \a statistic, suitable for logging.
 */
std::string ConsoleUtils::toString(const qpid::console::SchemaStatistic &statistic)
{
    return statistic.name + ':' + qmfTypeCodeToString(statistic.typeCode) + ':' +
           statistic.unit + ':' + statistic.desc;
}
